controllers/kustomization_controller.go

```diff
@@ -865,3 +843,77 @@ func (r *KustomizationReconciler) patchStatus(ctx context.Context, req ctrl.Requ
 
 	return r.Status().Patch(ctx, &kustomization, patch)
 }
+
+func (r *KustomizationReconciler) buildKubectlCmdImpersonation(ctx context.Context, kustomization kustomizev1.Kustomization, imp *KustomizeImpersonation, cmd string) (string, error) {
+	enableFluxUsers := r.EnableFluxUsers
+
+	if kustomization.Spec.ServiceAccountName != "" && kustomization.Spec.Principal == nil {
+		enableFluxUsers = false
+		kustomization.Spec.Principal = &kustomizev1.Principal{
+			Name: kustomization.Spec.ServiceAccountName,
+			Kind: "ServiceAccount",
+		}
+	}
+
+	if kustomization.Spec.KubeConfig != nil {
+		kubeConfig, err := imp.WriteKubeConfig(ctx)
+		if err != nil {
+			return "", err
+		}
+		cmd = fmt.Sprintf("%s --kubeconfig=%s", cmd, kubeConfig)
+	}
+
+	if enableFluxUsers {
+		var user string
+		if kustomization.Spec.Principal == nil {
+			user = fmt.Sprintf("flux:user:%s:%s", kustomization.Namespace, "reconciler")
+		} else if kustomization.Spec.Principal.Kind == "ServiceAccount" {
+			user = fmt.Sprintf("system:serviceaccount:%s:%s", kustomization.Namespace, kustomization.Spec.Principal.Name)
+		} else if kustomization.Spec.Principal.Kind == "User" {
+			user = fmt.Sprintf("flux:user:%s:%s", kustomization.Namespace, kustomization.Spec.Principal.Name)
+		} else {
+			return "", errors.New("unknown kind for impersonation, accepted kinds are User and ServiceAccount")
+		}
+
+		cmd = fmt.Sprintf("%s --as %s", cmd, user)
+
+		if c, err := rest.InClusterConfig(); err == nil && c != nil {
+			cmd = fmt.Sprintf("%s --certificate-authority=%s --token=%s --server=%s", cmd, c.CAFile, c.BearerToken, c.Host)
+		}
+
+		if kustomization.Spec.Principal != nil && kustomization.Spec.Principal.Kind == "ServiceAccount" {
+			return cmd, nil
+		}
+
+		cmd = fmt.Sprintf("%s --as-group flux:users --as-group flux:users:%s", cmd, kustomization.Namespace)
+		return cmd, nil
+	}
+
+	if kustomization.Spec.Principal == nil {
+		return cmd, nil
+	}
+
+	var saName string
+	if kustomization.Spec.ServiceAccountName != "" {
+		saName = kustomization.Spec.ServiceAccountName
+	} else if kustomization.Spec.Principal.Kind == "ServiceAccount" {
```

Comment:
Add the `ServiceAccount` and `User` strings to the runtime package as vars and use them in this PR please.